drivers/codec/wm8960.py

```diff
@@ -0,0 +1,783 @@
+            raise ValueError("Invalid input name")
+
+    def set_right_input(self, input):
+
+        regs = self.regs
+        if input == input_closed:
+            # Disable the input
+            regs[_POWER1] = (_POWER1_AINR_MASK | _POWER1_ADCR_MASK, 0)
+
+        elif input == input_mic1:
+            # Only LMN1 enabled, LMICBOOST to 13db, LMIC2B enabled
+            regs[_POWER1] = (0, _POWER1_AINR_MASK | _POWER1_ADCR_MASK | _POWER1_MICB_MASK)
+            regs[_RINPATH] = 0x138
+            regs[_RINVOL] = 0x117
+
+        elif input == input_mic2:
+            regs[_POWER1] = (0, _POWER1_AINR_MASK | _POWER1_ADCR_MASK | _POWER1_MICB_MASK)
+            regs[_RINPATH] = 0x178
+            regs[_RINVOL] = 0x117
+
+        elif input == input_mic3:
+            regs[_POWER1] = (0, _POWER1_AINR_MASK | _POWER1_ADCR_MASK | _POWER1_MICB_MASK)
+            regs[_RINPATH] = 0x1B8
+            regs[_RINVOL] = 0x117
+
+        elif input == input_line2:
+            regs[_POWER1] = (0, _POWER1_AINR_MASK | _POWER1_ADCR_MASK)
+            regs[_INBMIX2] = (0, 0xE)
+
+        elif input == input_line3:
+            regs[_POWER1] = (0, _POWER1_AINR_MASK | _POWER1_ADCR_MASK)
+            regs[_INBMIX2] = (0, 0x70)
+
+        else:
+            raise ValueError("Invalid input name")
+
+    def set_protocol(self, protocol):
+        self.regs[_IFACE1] = (
+            _IFACE1_FORMAT_MASK | _IFACE1_LRP_MASK,
+            protocol,
+        )
+
+    def config_data_format(self, sysclk, sample_rate, bits):
+        #  Compute sample rate divider, dac and adc are the same sample rate
+        try:
+            divider = self._dac_divider_table[sysclk // sample_rate]
+            wl = self._audio_word_length_table[bits]
+        except:
+            raise ValueError("Invalid ratio sysclk/sample_rate or invalid bit length")
+
+        self.regs[_CLOCK1] = (0x1F8, divider << 6 | divider << 3)
+        self.regs[_IFACE1] = (_IFACE1_WL_MASK, wl << _IFACE1_WL_SHIFT)
+
+    def set_volume(self, module, volume, volume_r=None):
+
+        if volume_r is None:
+            volume_r = volume
+
+        if not ((0 <= volume <= 100) and (0 <= volume_r <= 100)):
+            raise ValueError("Invalid value for volume")
+        elif not module in self._volume_config_table.keys():
+            raise ValueError("Invalid module")
+
+        vol_max, regnum, flags = self._volume_config_table[module]
+        self.regs[regnum] = int(volume * vol_max / 100 + 0.5) | flags
+        self.regs[regnum + 1] = int(volume_r * vol_max / 100 + 0.5) | flags
+
+    def get_volume(self, module):
+
+        if not module in self._volume_config_table.keys():
+            raise ValueError("Invalid module")
+
+        vol_max, regnum, _ = self._volume_config_table[module]
+        return (int((self.regs[regnum] & vol_max) * 100 / vol_max + 0.5),
+                int((self.regs[regnum + 1] & vol_max) * 100 / vol_max + 0.5))
+
+    def volume(self, module, volume_l=None, volume_r=None):
```

Comment:
This method is probably not needed?  Since it's just a shortcut for set/get.

Reply:
It prefer keeping the volume() method and drop set_volume() and get_volume().

Reply:
Done and pushed.